zne/noise_amplification/folding_amplifier/glorious_local_folding_amplifier.py

```diff
@@ -0,0 +1,170 @@
+        num_full_foldings = self._compute_num_foldings(noise_factor)
+        return self._apply_local_folding(dag, num_full_foldings, gates_to_fold)
+
+        # TODO: Find graceful way of implementing this without overriding typing
+        # TODO: Find way to work with standalone int and str instead of lists
+
+    def _apply_local_folding(
+        self,
+        dag: DAGCircuit,
+        num_foldings: int,
+        gates_to_fold: Sequence[int | str] | None,
+    ) -> DAGCircuit:
+        """Applies local folding strategy and returns a noise amplified dag.
+
+        Args:
+            dag (DAGCircuit): The original dag circuit without foldings.
+            num_foldings (float): Number of times the circuit should be folded.
+            gates_to_fold (Sequence[int | str] | None) : Original gates_to_fold input.
+
+        Returns:
+            DAGCircuit: The noise amplified DAG circuit.
+        """
+        self._validate_gates_to_fold(gates_to_fold)
+        if gates_to_fold in [0, "0", None]:
+            return dag
+        noisy_dag = dag.copy_empty_like()
+        for node in dag.topological_op_nodes():
+            if node.name in gates_to_fold or node.op.num_qubits in gates_to_fold:
+                noisy_dag = self._insert_folding(noisy_dag, node, num_foldings)
+            else:
+                noisy_dag.apply_operation_back(node.op, qargs=node.qargs, cargs=node.cargs)
+        return noisy_dag
+
+    def _insert_folding(
+        self,
+        dag: DAGCircuit,
+        node: DAGOpNode,
+        num_foldings: int,
+    ) -> DAGCircuit:
+        """Folds each gate of original DAG circuit a number of ``num_foldings`` times.
+
+        Args:
+            dag (DAGCircuit): The original dag circuit without foldings.
+            num_foldings (float): Number of times the circuit should be folded.
+
+        Returns:
+            DAGCircuit: The noise amplified DAG circuit.
+        """
+        barrier = Barrier(node.op.num_qubits)
+        dag.apply_operation_back(barrier, qargs=node.qargs)
+        dag.apply_operation_back(node.op, qargs=node.qargs, cargs=node.cargs)
+        inverted_node = node.op.inverse()
+        for _ in range(num_foldings):
+            dag.apply_operation_back(barrier, qargs=node.qargs)
+            dag.apply_operation_back(inverted_node, qargs=node.qargs, cargs=node.cargs)
+            dag.apply_operation_back(barrier, qargs=node.qargs)
+            dag.apply_operation_back(node.op, qargs=node.qargs, cargs=node.cargs)
+        dag.apply_operation_back(barrier, qargs=node.qargs)
+        return dag
+
+    def _compute_num_foldings(self, noise_factor: float) -> int:
+        """Compute number of foldings.
+
+        Args:
+            noise_factor (float) : The original noise_factor input.
+
+        Returns:
+            int: Number of foldings calculated from noise_factor.
+        """
+        noise_factor = self._validate_noise_factor(noise_factor)
+        return int((noise_factor - 1) / 2)
```

Comment:
We also had this function in the global amplifier, right? Any ideas of what we should do to avoid code repetition? 🔁